tests/test_callbacks.py

```diff
@@ -57,6 +63,19 @@ def test_callbacks(tmp_path, model_class):
     if os.path.exists(log_folder):
         shutil.rmtree(log_folder)
 
+    # Testing for models that support multiple envs
+    if model_class in [A2C, PPO]:
+        envs = make_vec_env(env_name, n_envs=2, seed=0, vec_env_cls=SubprocVecEnv)
+
+        model = model_class("MlpPolicy", envs, policy_kwargs=dict(net_arch=[32]))
+
+        callback = CallbackList([callback_max_eps])
+
+        model.learn(1000, callback=callback)
```

Comment:
Looks like should end with maximum episodes reached, being Pendulum-v0 envs with maximum 200 steps. Add check that this indeed happens.

Reply:
Saw no examples on how to check this automatically, and assumed that was by design, thus just verified behavior manually.

Reply:
@araffin Any comment?

Reply:
as @Miffyli suggests, please use `model.num_timesteps` to check how many episodes were run (because there is 200 timesteps per episode)

Reply:
Sure, the issue is that I am getting `model.num_timesteps= 398` when I ran multiple envs, and I am not sure why this is happening. I read the documentation about vecenvs and multiple environments, but still do not why this happens, is this an index issue? It is also odd that when I print the callback output it says `Stopping training with a total of 399 steps because`.

Reply:
this is normal as `self.num_timesteps += env.num_envs` is incremented after the `callback.on_step()`, I think you can those lines up, same for the off policy case

Reply:
Sorry, I do not understand. The suggestion was to add a code for checking `num_timesteps`, which I would propose something like this `assert (model.num_timesteps // n_envs) == 200`. Is this right? Where should I place it?

Reply:
@araffin I think I get it now. Do you want me to move up the line `self.num_timesteps += env.num_envs`, inside the `on_policy_algorithm.py` and `off_policy_algorithm.py` files?

Reply:
Yes that's how I understood. Personally I think best place is just after doing the calls to `step` functions.

Reply:
Ok, I will do it. However, the issue is that moving up the `self.num_timesteps += env.num_envs` might break some code. For example, the `on_step()` method in `BaseCallback` has a internal counter `self.num_timesteps = self.model.num_timesteps + 1`. Thus, I am not sure if this was a workaround given how `num_timesteps` is updated inside `on_policy_algorithm.py` and `off_policy_algorithm.py` base classes, or if there is another reason for having this internal counter.

In resume, do you want me to also change this line `self.num_timesteps = self.model.num_timesteps + 1` inside `callbacks.py` file?

Reply:
yes, please do (and check that the behavior is the same, for instance using a checkpoint callback)

Reply:
Sure, I will do it. But, looking closer at both policy base classes, why is this verification `if callback.on_step() is False:`happening so early on? Instead of moving up `self.num_timesteps += env.num_envs` shouldn't the callback check be moved down?

PS: Sorry for all the questions, but it is your project, and I really don't want to do undesired changes. =)